### ISSUE TITLE
Documentation - (Fancy) Boolean array indexing - incorrect sentence

### ISSUE THREAD

**Issue by meridianphage** [numpy/numpy#23377]
This section of the "Indexing on ndarrays" documentation page https://numpy.org/doc/stable/user/basics.indexing.html refers to indexing an array, x, with an object, obj. Unless I'm misreading something, the last sentence of the 2nd paragraph is incorrect. For example, doing x[obj] where x.shape=(4,3) and obj.shape=(4,2) gives an index error, which would not happen had you filled obj with **False**.

Also, the 2nd last sentence could improved - you get an index error if obj has values at entries that are outside of the bounds of x, regardless of whether those values are **True** or **False**.

![Documentation screenshot](https://user-images.githubusercontent.com/126125892/224602311-464072df-2f10-4d8c-8e7b-6d221a8248ae.png)

**Comment by RishabhSpark** [comment]
Hi! I'd like to work on this. I'm new to open-source and numpy, can someone please link me this in the code^?

**Comment by meridianphage** [comment]
@asmeurer Yes I'm not sure where it came from. The first paragraph claims
"A single boolean index array is practically identical to x[obj.nonzero()]"
but that's only true in general if obj.shape == x.shape , so that could've led to the confusion.

**Comment by asmeurer** [comment]
No that's true for any boolean array index (technically it isn't true when `obj` is 0-D, but that's an odd corner case). 

```py
>>> a = np.arange(12).reshape((4,3))
>>> obj = np.array([True, False, True, True])
>>> obj.shape
(4,)
>>> a[obj]
array([[ 0,  1,  2],
       [ 6,  7,  8],
       [ 9, 10, 11]])
>>> a[obj.nonzero()]
array([[ 0,  1,  2],
       [ 6,  7,  8],
       [ 9, 10, 11]])
```

**Comment by meridianphage** [comment]
(Yes please let's ignore 0-D cases haha)
I should have emphasized that I'm talking about what's true _in general_. There are many cases like the one you gave where the equivalence holds. But in your example, if obj had one fewer elements, or if you append an extra **False** element to the end, then a[obj.nonzero()] is fine but a[obj] gives an IndexError.

**Comment by asmeurer** [comment]
Oh I see your point. Yes, they are equivalent *if* the boolean array index is itself valid. But of course, *any* shaped integer array index is always valid, as long as the entries are not out of bounds for the axis size, so you're right that adding `False` to the end of a valid boolean index would be fine. It is indeed possible this is the source of the original confusion.

**Comment by seberg** [comment]
> Did boolean indexing used to work like this?

Yes, 10 years ago, the docs should have been exactl right.

**Comment by Unessam** [comment]
@asmeurer @meridianphage I can fix the issue, could you please confirm if my understanding of the issue is correct:
To fix the issue in the NumPy documentation, the second paragraph of the "Boolean array indexing" section should be revised to state that a boolean index array must have the same shape as the indexed array for the indexing operation to be valid. The last sentence should also be revised to clarify that an index error will be raised if obj has values at entries that are outside the bounds of x, regardless of whether those values are True or False.

**Comment by seberg** [comment]
@Unessam yes something like that, but it's one thing: The shape must match, and because of that there can't be out-of-bound indices to begin with (so there is no need to discuss that).

**Comment by asmeurer** [comment]
> a boolean index array must have the same shape as the indexed array for the indexing operation to be valid

This is not strictly true:

```py
>>> a = np.ones((2, 3, 4, 5))
>>> idx = np.ones((3, 4), dtype=bool)
>>> a[0, idx]
array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])
```

The more correct statement is that a boolean array must match a subset of the indexed array's shape in the corresponding location where it is being indexed. Matching the full array shape exactly is a special case of this, and is the most common one.

**Comment by Unessam** [comment]
@asmeurer thanks for the reply, does the below statement make sense based on the discussion?

It is important to note that a boolean index array must match a subset of the indexed array's shape in the corresponding location where it is being indexed. If obj.ndim == x.ndim, x[obj] returns a 1-dimensional array filled with the elements of x corresponding to the True values of obj. The search order will be row-major, C-style. An index error will be raised if obj has values at entries that are outside the bounds of x, regardless of whether those values are True or False.

**Comment by asmeurer** [comment]
That sounds correct, although "if obj has values at entries that are outside the bounds of x" is a little awkwardly stated, and perhaps misleading. I would rather say "if the shape `obj` does not match the corresponding dimensions of `x`". 

Also note that it's equally a problem for `obj` to be smaller than `x`. The size of the dimensions have to match exactly.

**Comment by Unessam** [comment]
@seberg could you please advise me on how to edit the user guide? it seems the generated text there is decoupled from the repository and I couldn't figure out any way to modify the text.

**Comment by charris** [comment]
Look in `doc/source/user/basics.indexing.rst`.

**Comment by Unessam** [comment]
@charris brilliant, thanks

**Comment by seberg** [comment]
Closing, the above PR should clarify most of it.
